src/Analyser/MutatingScope.php

```diff
@@ -3756,128 +3725,79 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 				continue;
 			}
-
-			$typeHolder = null;
-
+			$newConditionalExpressions[$variableExprString] = $conditionalExpressions;
 			foreach ($conditionalExpressions as $conditionalExpression) {
-				$matchingConditions = [];
 				foreach ($conditionalExpression->getConditionExpressionTypes() as $conditionExprString => $conditionalType) {
-					if (!array_key_exists($conditionExprString, $typeGuards)) {
-						continue;
-					}
-
-					if (!$typeGuards[$conditionExprString]->equals($conditionalType)) {
-						continue;
+					$conditionExpr = $this->exprStringToExpr($conditionExprString);
+					if ($conditionExpr === null) {
+						continue 2;
 					}
-
-					$matchingConditions[$conditionExprString] = $conditionalType;
-				}
-
-				if (count($matchingConditions) === 0) {
-					$newConditionalExpressions[$variableExprString][$conditionalExpression->getKey()] = $conditionalExpression;
-					continue;
-				}
-
-				if (count($matchingConditions) < count($conditionalExpression->getConditionExpressionTypes())) {
-					$filteredConditions = $conditionalExpression->getConditionExpressionTypes();
-					foreach (array_keys($matchingConditions) as $conditionExprString) {
-						unset($filteredConditions[$conditionExprString]);
+					if (!$scope->getType($conditionExpr)->equals($conditionalType)) {
+						continue 2;
 					}
-
-					$holder = new ConditionalExpressionHolder($filteredConditions, $conditionalExpression->getTypeHolder());
-					$newConditionalExpressions[$variableExprString][$holder->getKey()] = $holder;
-					continue;
 				}
 
 				$typeHolder = $conditionalExpression->getTypeHolder();
-				break;
-			}
-
-			if ($typeHolder === null) {
-				continue;
-			}
-
-			if ($typeHolder->getCertainty()->no()) {
-				unset($scope->expressionTypes[$variableExprString]);
-			} else {
-				$scope->expressionTypes[$variableExprString] = $typeHolder;
+				if ($typeHolder->getCertainty()->no()) {
+					unset($scope->expressionTypes[$variableExprString]);
+				} else {
+					$scope->expressionTypes[$variableExprString] = $typeHolder;
+				}
 			}
 		}
 
-		return $scope->changeConditionalExpressions($newConditionalExpressions);
+		return $scope->scopeFactory->create(
+			$scope->context,
+			$scope->isDeclareStrictTypes(),
+			$scope->getFunction(),
+			$scope->getNamespace(),
+			$scope->expressionTypes,
+			$newConditionalExpressions,
+			$scope->inClosureBindScopeClass,
+			$scope->anonymousFunctionReflection,
+			$scope->inFirstLevelStatement,
+			$scope->currentlyAssignedExpressions,
+			$scope->currentlyAllowedUndefinedExpressions,
+			$scope->nativeExpressionTypes,
+			$scope->inFunctionCallsStack,
+			$scope->afterExtractCall,
+			$scope->parentScope,
+		);
 	}
 
-	/**
-	 * @param array<string, ConditionalExpressionHolder[]> $newConditionalExpressionHolders
-	 */
-	public function changeConditionalExpressions(array $newConditionalExpressionHolders): self
+	private function exprStringToExpr(string $exprString): ?Expr
```

Comment:
This is added temporary. Could be deleted after ConditionalExpressionHolder condition types is changed to ExpressionTypeHolders

Reply:
Yeah, I'd welcome that :)